src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -961,25 +947,87 @@ bool tryProcessTree(SyntaxTree partialTree, [NotNullWhen(true)] out (AnalysisSco
 
             return dequeueAndFilterCompilationEvents(compilation, analysisScope, additionalFiles, cancellationToken);
 
-            static void generateCompilationEvents(Compilation compilation, AnalysisScope analysisScope, CancellationToken cancellationToken)
+            static async Task generateCompilationEventsAsync(Compilation compilation, AnalysisScope analysisScope, ConcurrentSet<ISymbol>? addedSymbols, ConcurrentSet<SyntaxTree>? addedTrees, CancellationToken cancellationToken)
             {
+                Debug.Assert(!analysisScope.IsEntireCompilationAnalysis);
+                Debug.Assert(!analysisScope.IsSyntacticSingleFileAnalysis);
+
                 // Invoke GetDiagnostics to populate CompilationEvent queue for the given analysis scope.
                 // Discard the returned diagnostics.
-                if (analysisScope.FilterFileOpt == null)
-                {
-                    _ = compilation.GetDiagnostics(cancellationToken);
-                }
-                else if (!analysisScope.IsSyntacticSingleFileAnalysis)
+                if (analysisScope.FilterFileOpt.HasValue)
                 {
                     // Get the mapped model and invoke GetDiagnostics for the given filter span, if any.
                     // Limiting the GetDiagnostics scope to the filter span ensures we only generate compilation events
                     // for the required symbols whose declaration intersects with this span, instead of all symbols in the tree.
-                    var mappedModel = compilation.GetSemanticModel(analysisScope.FilterFileOpt!.Value.SourceTree!);
-                    _ = mappedModel.GetDiagnostics(analysisScope.FilterSpanOpt, cancellationToken);
+                    generateCompilationEventsForTree(analysisScope.FilterFileOpt.Value.SourceTree!);
+                }
+                else
+                {
+                    // We have a multiple file analysis scope, which shouldn't have a filter span.
+                    Debug.Assert(!analysisScope.FilterSpanOpt.HasValue);
+
+                    // Invoke GetDiagnostics for all the files in the analysis scope.
+                    if (analysisScope.ConcurrentAnalysis)
+                    {
+                        await Task.WhenAll(analysisScope.SyntaxTrees.Select(tree =>
+                            Task.Run(() =>
+                            {
+                                generateCompilationEventsForTree(tree);
+                            }, cancellationToken))).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        foreach (var tree in analysisScope.SyntaxTrees)
+                        {
+                            generateCompilationEventsForTree(tree);
+                        }
+                    }
+                }
+
+                void generateCompilationEventsForTree(SyntaxTree tree)
```

Comment:
This is core change to synthesize compilation events for top level symbols in the file by walking the tree, instead of invoking `SemanticModel.GetDiagnostics()`